### PR TITLE
Updates the pre-rotation and witnesses sections with feedback from the first set of updates

### PR DESCRIPTION
**Pull request by swcurran**
Adjusts the recently changes pre-rotation and witness section, addressing a number of issues:

- Issue #151 adds that pre-rotation can be turned off, how (empty list) and the implications
- Issue #152 updates the `did-witness.json` data model to make the `versionId` what is signed (also issue #142)
- Issue #141 #143 #144 clarifying language on when and why to remove proofs and when there will be 1 or 2 proofs from a witness. Tricky language!
- Issue #149 the type of Data Integrity cryptosuite that the witnesss must use
- Issue #140 how to disable the use of witnesses for a DID

The goal is that this PR should complete the changes needed in v0.5 -- including the renaming.

Please review carefully!

### CODE REVIEW

File: spec/specification.md [hunk 1]
```diff
@@ -428,6 +428,24 @@ Document the full list of error codes that can be generated in resolving a DID.
 
 ##### Reading did:webvh DID URLs
 
+A `didLwebvh` resolver **MUST** resolve the [[spec:DID-Core]] `versionId` and
```
Comment (brianorwhatever) [inline spec/specification.md:431]:
```suggestion
A `did:webvh` resolver **MUST** resolve the [[spec:DID-Core]] `versionId` and
```

File: spec/specification.md [hunk 2]
```diff
@@ -964,10 +1001,9 @@ where:
 The use of the [[ref: threshold]] and weighted approvals (versus needing
 approvals from all [[ref: witnesses]]) is to prevent faulty [[ref: witnesses]]
 from blocking the publishing of a new version of the DID. To determine if the
-[[ref: threshold]] has been met, sum the `weight` integer of the received
-approvals and if it equal to or more than the `threshold`, the update can be
-published. The calculation **MUST**  be executed by resolvers processing a
-[[ref: DID Log]].
+[[ref: threshold]] has been met, all participants **MUST** sum the `weight`
+integer of the received approvals. and if it equal to or more than the
```
Comment (brianorwhatever) [inline spec/specification.md:1005]:
```suggestion
integer of the received approvals. and if it is equal to or more than the
```

File: spec/specification.md [hunk 3]
```diff
@@ -976,40 +1012,60 @@ all of the first three witness (`1+1+1`) providing an approving `proof`.
...
-Proofs from witnesses are retained in a separate file (`did-witness.json`) from the
-[[ref: DID Log]]. The same [DID to HTTPS
+Proofs from [[ref: witnesses]] are placed into a separate file
+(`did-witness.json`) from the [[ref: DID Log]]. The same [DID to HTTPS
 Transformation](#the-did-to-https-transformation) used for the [[ref: DID Log]]
 is used to locate the `did-witness.json` resource, with only the last element
-changed (`did.jsonl` to `did-witness.json`). The media type of the file **SHOULD** be
-`text/json`.
+changed (`did.jsonl` to `did-witness.json`). The media type of the file
+**SHOULD** be `text/json`.
```
Comment (brianorwhatever) [inline spec/specification.md:1020]:
```suggestion
**SHOULD** be `application/json`.
```

File: spec/specification.md [hunk 4]
```diff
@@ -1019,23 +1075,20 @@ The following process is used to witness a DID version update:
   `proof` element) for the new version of the DID, and shares it with the active [[ref: witnesses]].
   - The specification leaves to implementers *how* the [[ref: log entry]] data is provided to the [[ref: witnesses]].
 - The [[ref: witnesses]] ***MUST** hold their own copy of the published [[ref:
-  DID Log]] prior to the version being witnessed.
+  DID Log]] prior witnessing a [[ref: DID Log entry]].
```
Comment (brianorwhatever) [inline spec/specification.md:1078]:
```suggestion
  DID Log]] prior to witnessing a [[ref: DID Log entry]].
```

File: spec/specification.md [hunk 5]
```diff
@@ -1019,23 +1075,20 @@ The following process is used to witness a DID version update:
...
-   as defined in the [Authorized Keys](#authorized-keys) section of this
-   specification. 
-- The DID Controller updates the [[ref: witnesses]]'s proofs in the
-  `did-witness.json` file with the received proofs, and publishes the updated file.
+  creates and sends to the [[ref: DID Controller]] a [[ref: Data Integrity]]
+  proof signed using the [[ref: witness]]'s `did:key` DID.
   - The specification leaves to implementers how [[ref: witness]] proofs are
     conveyed to the [[ref: DID Controller]].
+- The [[ref: DID Controller]] **MUST** create or add the proof to the record for
+  the applicable `versionID` for the unpublished [[ref: DID log entry]].
```
Comment (brianorwhatever) [inline spec/specification.md:1090]:
```suggestion
  the applicable `versionId` for the unpublished [[ref: DID log entry]].
```

File: spec/specification.md (hunk 3 above)
Comment (swcurran) [reply spec/specification.md]: I have the JSON Lines type as `text/jsonl`.  That's correct?

Comment (brianorwhatever) [reply spec/specification.md]: It's "correct" in that its the leading candidate. JSON of course has more eyeballs so this is what wikipedia says

The official [MIME type](https://en.wikipedia.org/wiki/MIME_type) for JSON text is application/json,[[26]](https://en.wikipedia.org/wiki/JSON#cite_note-27) and most modern implementations have adopted this. Legacy MIME types include text/json, text/x-json, and text/javascript.[[27]](https://en.wikipedia.org/wiki/JSON#cite_note-28) The standard filename extension is .json.[[28]](https://en.wikipedia.org/wiki/JSON#cite_note-29)

**Comment by brianorwhatever** [comment]
So with the new `null` option for nextKeyHashes what about `undefined` aka the parameter is left off. This should definitely be a failure case right?

**Comment by andrewwhitehead** [comment]
> So with the new `null` option for nextKeyHashes what about `undefined` aka the parameter is left off. This should definitely be a failure case right?

If a parameter is left off it should carry on with its old value, although in this case the update would be invalid because we require it to be changed.

My perspective was that parameters that have never been set default to `null` and they can be reset to `null` (since `undefined` isn't supported in JSON). Although we could also define default (implicit) values for each parameter.

**Comment by swcurran** [comment]
Need to get to a final answer on this.

1. We could add to the spec that all parameters are initialized to null.
2. We could add the default to null **and** remove the `empty list “[]”` option for the array items, leaving only `null` — used initially and when reset.

As noted in the spec, a missing value in the parameters means “unchanged from the previous value”. As Andrew notes,  some parameters add rules of when that a missing value is unacceptable — e.g. parameters that **MUST** be in the first entry, and the pre-rotation parameters that **MUST** be in every entry when active.

After writing this, my leaning is 2 above.

**Comment by swcurran** [comment]
Added commit to limit array parameters (`witness`, `nextKeyHashes`, `updateKeys`) to be deactivated by `null`, removing the option of using of empty list `[]`.  Also went through the parameters and formalized their default values where not previously explicit, and how to use `null` to deactive them (where needed).

**Comment by swcurran** [comment]
Add a last tweak -- made explicit that they array parameters **MUST NOT** be set to an empty list.

File: spec/specification.md [hunk 6]
```diff
@@ -597,11 +621,15 @@ properties are defined below.
...
+    [deactivate](#deactivate-revoke) section of this specification for more
+    details.
+  - `updateKeys` **MUST NOT** be set to an empty list `[]`.
 - `portable`: A boolean (`true` / `false`) indicating if the DID is portable and
   thus can be renamed to change the Web location of the DID.
   - The value can **ONLY** be set to `true` in the first [[ref: log entry]], the initial version of the DID.
-  - If not explicitly set in the first [[ref: log entry]], it is set to `false`.
-  - Once the value has been set to `false`, it cannot be set back to `true`.
+  - If not explicitly set in the first [[ref: log entry]], it **MUST** be set to `false`.
+  - Once the value has been explicitly to `false` in a [[ref: DID log entry]], it **MUST NOT** be set back to `true`.
```
Comment (brianorwhatever) [inline spec/specification.md:632]:
```suggestion
  - Once the value has been explicitly set to `false` in a [[ref: DID log entry]], it **MUST NOT** be set back to `true`.
```

File: spec/specification.md [hunk 7]
```diff
@@ -597,11 +621,15 @@ properties are defined below.
...
+  - `updateKeys` **SHOULD** be set to `null` when deactivating the DID. See the
+    [deactivate](#deactivate-revoke) section of this specification for more
+    details.
+  - `updateKeys` **MUST NOT** be set to an empty list `[]`.
 - `portable`: A boolean (`true` / `false`) indicating if the DID is portable and
   thus can be renamed to change the Web location of the DID.
   - The value can **ONLY** be set to `true` in the first [[ref: log entry]], the initial version of the DID.
-  - If not explicitly set in the first [[ref: log entry]], it is set to `false`.
-  - Once the value has been set to `false`, it cannot be set back to `true`.
+  - If not explicitly set in the first [[ref: log entry]], it **MUST** be set to `false`.
```
Comment (brianorwhatever) [inline spec/specification.md:631]: I think this kind of reads funny as it's not really a MUST it is just a fact.
```suggestion
  - If not explicitly set in the first [[ref: log entry]], its value is set to `false`.
```

File: spec/specification.md [hunk 8]
```diff
@@ -610,12 +638,21 @@ properties are defined below.
   - The process for generating the hashes and additional details for using [[ref: pre-rotation]] are defined in the
     [Pre-Rotation Key Hash Generation and Verification](#pre-rotation-key-hash-generation-and-verification)
     section of this specification.
-  - Once the [[ref: parameter]] `nextKeyHashes` has been set, the [[ref: Key Pre-Rotation]] feature becomes active and the property **MUST** be present in any future [[ref: DID log entry]].
+  - If not explicitly set in the first [[ref: DID Log entry]], its value **MUST** be `null`.
```
Comment (brianorwhatever) [inline spec/specification.md:641]: Same as above

```suggestion
  - If not explicitly set in the first [[ref: DID Log entry]], its value is `null`.
```

File: spec/specification.md [hunk 9]
```diff
@@ -610,12 +638,21 @@ properties are defined below.
...
   - All [[ref: multikey]] formatted public keys added in a new `updateKeys` list **MUST** have their
     hashes listed in the `nextKeyHashes` list from the previous [[ref: DID log entry]].
   - A [[ref: DID Controller]] **MAY** put extra strings in the `nextKeyHashes`
     array that are not subsequently used in an `updateKeys` entry.
   - Any unused hashes in the prior `nextKeyHashes` are ignored.
+  - The value of `nextKeyHashes` **MAY** be set to `null` to deactivate
+    pre-rotation. For additional details about turning off [[ref: pre-rotation]]
+    see the [pre-rotation](#pre-rotation-key-hash-generation-and-verification)
+    section of this specification.
+  - `nextKeyHashes` **MUST NOT** be set to an empty list `[]`.
```
Comment (brianorwhatever) [inline spec/specification.md:655]:
```suggestion
  - `nextKeyHashes` **MUST** have at least 1 entry and **MUST NOT** be set to an empty list `[]`.
```

File: spec/specification.md [hunk 10]
```diff
@@ -627,14 +664,19 @@ properties are defined below.
     a `witness` property becomes active **after** the publication of its entry
     -- meaning its [[ref: log entry]] **MUST** be witnessed by active
     `witnesses` from a **prior** [[ref: DID log]] entry.
-- `deactivated`: A JSON boolean that **SHOULD** be set to `true` when the DID is to
-  be deactivated. See the [deactivate (revoke)](#deactivate-revoke) section of
-  this specification for more details.
+  - If the `witness` property is not set in the first [[ref: DID log entry]],
+    its value **MUST** be null.
```
Comment (brianorwhatever) [inline spec/specification.md:668]:
```suggestion
    its value is null.
```

File: spec/specification.md [hunk 11]
```diff
@@ -627,14 +664,19 @@ properties are defined below.
...
+  - `witness` **MUST NOT** be set to an empty list `[]`.
```
Comment (brianorwhatever) [inline spec/specification.md:669]:
```suggestion
  - `witness` **MUST** have at least 1 entry and **MUST NOT** be set to an empty list `[]`.
```

File: spec/specification.md [hunk 12]
```diff
@@ -627,14 +664,19 @@ properties are defined below.
...
+- `deactivated`: A JSON boolean that **MUST** be initialized to `false` and
```
Comment (brianorwhatever) [inline spec/specification.md:670]:
```suggestion
- `deactivated`: A JSON boolean that is initialized to `false` and
```

File: spec/specification.md [hunk 13]
```diff
@@ -627,14 +664,19 @@ properties are defined below.
...
+  **SHOULD** be set to `true` when the DID is to be deactivated but remains
```
Comment (brianorwhatever) [inline spec/specification.md:671]:
```suggestion
  **MUST** be set to `true` when the DID is to be deactivated but remains
```

**Review by brianorwhatever — DISMISSED** [review]
All of my comments are of the  "take it or leave it" type

File: spec/specification.md (hunk 6 above)
Comment (swcurran) [reply spec/specification.md:632]: Fixed

File: spec/specification.md (hunk 7 above)
Comment (swcurran) [reply spec/specification.md:631]: I think leaving it as being testable makes sense.  Leaving this one.

File: spec/specification.md (hunk 8 above)
Comment (swcurran) [reply spec/specification.md:641]: No change

File: spec/specification.md (hunk 9 above)
Comment (swcurran) [reply spec/specification.md:655]: Fixed in all three places.

File: spec/specification.md (hunk 10 above)
Comment (swcurran) [reply spec/specification.md:668]: No change

File: spec/specification.md (hunk 11 above)
Comment (swcurran) [reply spec/specification.md:669]: Fixed

File: spec/specification.md (hunk 12 above)
Comment (swcurran) [reply spec/specification.md:670]: No change

File: spec/specification.md (hunk 13 above)
Comment (swcurran) [reply spec/specification.md:671]: Left as is. In the deactivate section we left it.  If a DID is "deactivated" by removing the JSONL file, for example, no need to set deactivate to true.  I wanted to leave that open.

**Review by brianorwhatever — APPROVED** [review]

**Pull request merged**